nav2_map_server/src/vo_server/vector_object_server.cpp

```diff
@@ -0,0 +1,582 @@
+        !nav2_util::worldToMap(map_, wx2, wy2, mx2, my2))
+      {
+        RCLCPP_ERROR(
+          get_logger(),
+          "Error to get shape boundaries on map (UUID: %s)",
+          shape->getUUID().c_str());
+        return;
+      }
+
+      unsigned int it;
+      for (unsigned int my = my1; my <= my2; my++) {
+        for (unsigned int mx = mx1; mx <= mx2; mx++) {
+          it = my * map_->info.width + mx;
+          double wx, wy;
+          nav2_util::mapToWorld(map_, mx, my, wx, wy);
+          if (shape->isPointInside(wx, wy)) {
+            processVal(map_->data[it], shape->getValue(), overlay_type_);
+          }
+        }
+      }
+    } else {
+      // Put shape borders on map
+      shape->putBorders(map_, overlay_type_);
+    }
+  }
+}
+
+void VectorObjectServer::publishMap()
+{
+  if (map_) {
+    auto map = std::make_unique<nav_msgs::msg::OccupancyGrid>(*map_);
+    map_pub_->publish(std::move(map));
+  }
+}
+
+void VectorObjectServer::processMap()
+{
+  if (!process_map_) {
+    return;
+  }
+
+  try {
+    if (shapes_.size()) {
+      if (!transformVectorObjects()) {
+        return;
+      }
+      double min_x, min_y, max_x, max_y;
+
+      getMapBoundaries(min_x, min_y, max_x, max_y);
+      updateMap(min_x, min_y, max_x, max_y);
+      putVectorObjectsOnMap();
+    } else {
+      updateMap(0.0, 0.0, 0.0, 0.0);
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Can not upate map: %s", ex.what());
+    return;
+  }
+
+  publishMap();
+}
+
+bool VectorObjectServer::isMapUpdate()
+{
+  for (auto shape : shapes_) {
+    if (shape->getFrameID() != frame_id_ && !shape->getFrameID().empty()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void VectorObjectServer::switchMapUpdate()
+{
+  if (isMapUpdate()) {
+    if (!map_timer_) {
+      map_timer_ = this->create_timer(
+        std::chrono::duration<double>(1.0 / update_frequency_),
+        std::bind(&VectorObjectServer::processMap, this));
+      RCLCPP_INFO(get_logger(), "Publishing map dynamically");
+    }
+  } else {
+    if (map_timer_) {
+      map_timer_->cancel();
+      map_timer_.reset();
+    }
+    RCLCPP_INFO(get_logger(), "Publishing map once");
+    processMap();
+  }
+}
+
+void VectorObjectServer::addShapesCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::AddShapes::Request> request,
+  std::shared_ptr<nav2_msgs::srv::AddShapes::Response> response)
+{
+  // Initialize result with true. If one of the required vector object was not added properly,
+  // set it to false.
+  response->success = true;
+
+  auto node = shared_from_this();
+
+  // Process polygons
+  for (auto req_poly : request->polygons) {
+    nav2_msgs::msg::PolygonVO::SharedPtr new_params =
+      std::make_shared<nav2_msgs::msg::PolygonVO>(req_poly);
+
+    bool new_shape = true;  // Whether to add a new shape
+    std::shared_ptr<Polygon> polygon;
+    for (auto shape : shapes_) {
```

Comment:
If the shapes were stored in an unordered map hashed by the uuid, then this would be a trivial lookup

Reply:
Nice item for optimization. Thank you for pointing this out! Made `shapes_` to be `std::unordered_map` in https://github.com/ros-planning/navigation2/commit/d6a831d90d1440436538a61c771f017bddb59422